controllers/provider-openstack/pkg/controller/infrastructure/actuator_reconcile.go

```diff
@@ -0,0 +1,188 @@
+	if err := a.client.Get(ctx, kutil.Key(infrastructure.Spec.SecretRef.Namespace, infrastructure.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	terraformConfig, err := generateTerraformInfraConfig(ctx, infrastructure, infrastructureConfig, cluster, extractCredentials(providerSecret))
+	if err != nil {
+		return fmt.Errorf("failed to generate Terraform config: %+v", err)
+	}
+
+	chartRenderer, err := chartrenderer.NewForConfig(a.restConfig)
+	if err != nil {
+		return fmt.Errorf("could not create chart renderer: %+v", err)
+	}
+
+	release, err := chartRenderer.Render(filepath.Join(openstack.InternalChartsPath, "openstack-infra"), "openstack-infra", infrastructure.Namespace, terraformConfig)
+	if err != nil {
+		return fmt.Errorf("could not render Terraform chart: %+v", err)
+	}
+
+	tf, err := a.newTerraformer(openstack.TerrformerPurposeInfra, infrastructure.Namespace, infrastructure.Name)
+	if err != nil {
+		return fmt.Errorf("could not create terraformer object: %+v", err)
+	}
+
+	if err := tf.
+		SetVariablesEnvironment(generateTerraformInfraVariablesEnvironment(providerSecret)).
+		InitializeWith(terraformer.DefaultInitializer(
+			a.client,
+			release.FileContent("main.tf"),
+			release.FileContent("variables.tf"),
+			[]byte(release.FileContent("terraform.tfvars"))),
+		).
+		Apply(); err != nil {
+
+		return &controllererrors.RequeueAfterError{
+			Cause:        err,
+			RequeueAfter: 30 * time.Second,
+		}
+	}
+
+	if err := a.updateProviderStatus(ctx, tf, infrastructure, infrastructureConfig); err != nil {
+		return fmt.Errorf("failed to update the provider status in the Infrastructure resource: %+v", err)
+	}
+	return nil
+}
+
+func (a *actuator) updateProviderStatus(ctx context.Context, tf *terraformer.Terraformer, infrastructure *extensionsv1alpha1.Infrastructure, infrastructureConfig *openstackv1alpha1.InfrastructureConfig) error {
+	outputVarKeys := []string{
+		openstack.SSHKeyName,
+		openstack.RouterID,
+		openstack.NetworkID,
+		openstack.SubnetID,
+		openstack.FloatingNetworkID,
+		openstack.SecurityGroupID,
+		openstack.SecurityGroupName,
+	}
+
+	output, err := tf.GetStateOutputVariables(outputVarKeys...)
+	if err != nil {
+		return err
+	}
+
+	infrastructure.Status.ProviderStatus = &runtime.RawExtension{
+		Object: &openstackv1alpha1.InfrastructureStatus{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: openstackv1alpha1.SchemeGroupVersion.String(),
+				Kind:       "InfrastructureStatus",
+			},
+			Router: openstackv1alpha1.RouterStatus{
+				ID: output[openstack.RouterID],
+			},
+			Network: openstackv1alpha1.NetworkStatus{
+				ID: output[openstack.NetworkID],
+				Subnets: []openstackv1alpha1.Subnet{
+					{
+						ID:      output[openstack.SubnetID],
+						Purpose: openstackv1alpha1.PurposeNodes,
+					},
+				},
+			},
+			Node: openstackv1alpha1.NodeStatus{
+				KeyName: output[openstack.SSHKeyName],
+			},
+		},
+	}
+
+	return a.client.Status().Update(ctx, infrastructure)
```

Comment:
Similar to https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/pkg/controller/infrastructure/actuator_reconcile.go#L192 this should be retried I guess